In [3]:
import nltk
from nltk.corpus import wordnet

<b>1) Найти все значения (синсеты) для лексемы plant<b>

In [4]:
play = wordnet.synsets('play')
for ss in play:
    print(ss, ss.definition())

Synset('play.n.01') a dramatic work intended for performance by actors on a stage
Synset('play.n.02') a theatrical performance of a drama
Synset('play.n.03') a preset plan of action in team sports
Synset('maneuver.n.03') a deliberate coordinated movement requiring dexterity and skill
Synset('play.n.05') a state in which action is feasible
Synset('play.n.06') utilization or exercise
Synset('bid.n.02') an attempt to get something
Synset('play.n.08') activity by children that is guided more by imagination than by fixed rules
Synset('playing_period.n.01') (in games or plays or other performances) the time during which play proceeds
Synset('free_rein.n.01') the removal of constraints
Synset('shimmer.n.01') a weak and tremulous light
Synset('fun.n.02') verbal wit or mockery (often at another's expense but not to be taken seriously)
Synset('looseness.n.05') movement or space for movement
Synset('play.n.14') gay or light-hearted recreational activity for diversion or amusement
Synset('turn.n.0

<b>2) Найти определение для лексемы play в значении (а) "спектакль" и в значении (b) "играть в игры"</b>

In [18]:
print(play[0], play[0].definition())
print(play[17], play[17].definition())

Synset('play.n.01') a dramatic work intended for performance by actors on a stage
Synset('play.v.01') participate in games or sport


<b>3) Найдите два произвольных контекста для слова plant в значениях (a) "спектакль" и (b) "играть в игры"; продемонстрируйте на них действие алгоритма Леска для разрешения неоднозначности</b>

In [19]:
sent1 = 'Let’s play a game that called Alias.'
sent2 = 'The play was written by Jeroen Olyslaegers and is directed by Floor Huygen.'
sent1_tokens = [word.strip('.,') for word in sent1.split(' ')]
sent2_tokens = [word.strip('.,') for word in sent2.split(' ')]
print(sent1_tokens)
print(sent2_tokens)

['Let’s', 'play', 'a', 'game', 'that', 'called', 'Alias']
['The', 'play', 'was', 'written', 'by', 'Jeroen', 'Olyslaegers', 'and', 'is', 'directed', 'by', 'Floor', 'Huygen']


In [20]:
from nltk.wsd import lesk
print(lesk(sent1_tokens, 'play').definition())
print(lesk(sent2_tokens, 'play').definition())

put (a card or piece) into play during a game, or act strategically as if in a card game
activity by children that is guided more by imagination than by fixed rules


Алгоритм Леска ошибся как в первом, так и во втором случае, хотя часть речи везде была определена правильно.

<b>4) Найдите гиперонимы для значения (a) и гиперонимы для значения (b)</b>

In [21]:
for ss in play[0].hypernyms():
    print(ss, ss.definition())
for ss in play[17].hypernyms():
    print(ss, ss.definition())

Synset('dramatic_composition.n.01') a play for performance on the stage or television or in a movie etc.
Synset('compete.v.01') compete for something; engage in a contest; measure oneself against others


<b>5) Вычислите наименьшее расстояние между значением play "спектакль" и значениями лексемы theatre, а также play "играть в игры" и значениями лексемы game.
Найти min (d(plant: "спектакль", call), d(plant: "спектакль", game)), а также min (d(plant: "играть в игры", call), d(plant: "играть в игры", game))</b>

In [35]:
call = wordnet.synsets('call')
game = wordnet.synsets('game')

def get_dist_sim(ss1, lexeme):
    distances = []
    similarities = []
    for ss2 in lexeme:
        dist = ss1.shortest_path_distance(ss2)
        if dist is not None:
            distances.append(dist)
            sim = ss1.path_similarity(ss2)
            similarities.append(sim)
    return distances, similarities

# min d(play: "спектакль", call)
dist1 = get_dist_sim(play[0], call)[0]
print('min d(play: "cпектакль", call): {}'.format(min(dist1)))
print('closest lemma definition: {}\n'.format(call[dist1.index(min(dist1))].definition()))

# min d(play: "спектакль", game)
dist2 = get_dist_sim(play[0], game)[0]
print('min d(play: "спектакль", game): {}'.format(min(dist2)))
print('closest lemma definition: {}\n'.format(game[dist2.index(min(dist2))].definition()))

# min d(play: "играть в игры", call)
dist3 = get_dist_sim(play[17], call)[0]
print('min d(play: "играть в игры", call): {}'.format(min(dist3)))
print('closest lemma definition: {}\n'.format(call[dist3.index(min(dist3))].definition()))

# min d(play: "играть в игры", game)
dist4 = get_dist_sim(play[17], game)[0]
print('min d(play: "играть в игры", game): {}'.format(min(dist4)))
print('closest lemma definition: {}\n'.format(game[dist4.index(min(dist4))].definition()))

print('min (d(play: "спектакль", call), d(play: "спектакль", game)): {}'.format(min(min(dist1), min(dist2))))
print('min (d(play: "играть в игры", call), d(play: "играть в игры", game)): {}'.format(min(min(dist3), min(dist4))))

min d(play: "cпектакль", call): 7
closest lemma definition: a loud utterance; often in protest or opposition

min d(play: "спектакль", game): 10
closest lemma definition: a contest with rules to determine a winner

min d(play: "играть в игры", call): 1
closest lemma definition: a telephone connection

min d(play: "играть в игры", game): 3
closest lemma definition: a contest with rules to determine a winner

min (d(play: "спектакль", call), d(play: "спектакль", game)): 7
min (d(play: "играть в игры", call), d(play: "играть в игры", game)): 1


Кратчайшее расстояние от спектакля до call - 7 (у ближайшей леммы call не совсем ожидаемое для данного задания значение), до game - 10. 
Кратчайшее расстояние от растения до call - 1, до game - 3. До game расстояние оказалось длиннее, что было не совсем ожидаемо. Казалось, что game ближе к play в значении "играть в игры".



<b>6) Вычислить двумя разными способами расстояние:
d(play: "спектакль", comedy) и d(organism, whole)
Есть ли разница в расстояниях? Какое из расстояний, по Вашему мнению, в лучшей степени отражает интуитивное представление о семантчиеской близости слов?</b>


In [39]:
com = wordnet.synsets("comedy")
for ss in com:
    print(ss, ss.definition())

Synset('comedy.n.01') light and humorous drama with a happy ending
Synset('drollery.n.01') a comic incident or series of incidents


Посчитаем близость с помощью критериев Path Similarity, Leacock-Chodorow Similarity, Wu-Palmer Similarity, Resnik Similarity, Jiang-Conrath Similarity и Lin Similarity. Для последних трех будем использовать Information Content корпуса Brown.

In [42]:
from nltk.corpus import wordnet_ic
ic = wordnet_ic.ic('ic-brown.dat')

print(play[0].path_similarity(com[0]))

#Leacock-Chodorow Similarity
print(play[0].lch_similarity(com[0]))

#Wu-Palmer Similarity
print(play[0].wup_similarity(com[0]))
    
#Resnik Similarity
print(play[0].res_similarity(com[0], ic))
    
#Jiang-Conrath Similarity
print(play[0].jcn_similarity(com[0], ic))
    
#Lin Similarity
print(play[0].lin_similarity(com[0], ic))


0.1111111111111111
1.4403615823901665
0.42857142857142855
2.61964384137646
0.07125094921216772
0.2718292239944514


In [43]:
organism = wordnet.synsets('organism')
for ss in organism:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('organism.n.01') a living thing that has (or can develop) the ability to act or function independently
Synset('organism.n.01') []
Synset('organism.n.02') a system considered analogous in structure or function to a living body
Synset('organism.n.02') ['the social organism']


In [44]:
whole = wordnet.synsets('whole', 'n')
for ss in whole:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('whole.n.01') all of something including all its component elements or parts
Synset('whole.n.01') ['Europe considered as a whole', 'the whole of American literature']
Synset('whole.n.02') an assemblage of parts that is regarded as a single entity
Synset('whole.n.02') ['how big is that part compared to the whole?', 'the team is a unit']


In [45]:
for ss1 in organism:
    print(ss1.definition())
    for ss2 in whole:
        print(ss2.definition())
        print(ss1.path_similarity(ss2))

        #Leacock-Chodorow Similarity
        print(ss1.lch_similarity(ss2))

        #Wu-Palmer Similarity
        print(ss1.wup_similarity(ss2))

        #Resnik Similarity
        print(ss1.res_similarity(ss2, ic))

        #Jiang-Conrath Similarity
        print(ss1.jcn_similarity(ss2, ic))

        #Lin Similarity
        print(ss1.lin_similarity(ss2, ic))

a living thing that has (or can develop) the ability to act or function independently
all of something including all its component elements or parts
0.07692307692307693
1.072636802264849
0.14285714285714285
-0.0
0.11927453545173074
-0.0
an assemblage of parts that is regarded as a single entity
0.3333333333333333
2.538973871058276
0.8
1.5318337432196856
1.4444255924181877
0.8156763478002892
a system considered analogous in structure or function to a living body
all of something including all its component elements or parts
0.1
1.3350010667323402
0.3076923076923077
0.5962292078977726
0.05928363701403374
0.0660256991738309
an assemblage of parts that is regarded as a single entity
0.125
1.55814461804655
0.2222222222222222
-0.0
0.07444639262208605
-0.0


<style type="text/css">
	table.tableizer-table {
		font-size: 12px;
		border: 1px solid #CCC; 
		font-family: Arial, Helvetica, sans-serif;
	} 
	.tableizer-table td {
		padding: 4px;
		margin: 3px;
		border: 1px solid #CCC;
	}
	.tableizer-table th {
		background-color: #104E8B; 
		color: #FFF;
		font-weight: bold;
	}
</style>
<table class="tableizer-table">
<thead><tr class="tableizer-firstrow"><th>Similarity Index</th><th>спектакль/comedy</th><th>a living thing / all of something</th><th>a living thing / an assemblage</th><th>a system / all of something</th><th>a system / an assemblage</th></tr></thead><tbody>
 <tr><td>Path Similarity</td><td>0.1111111111</td><td>0.07692307692</td><td>0.3333333333</td><td>0.1</td><td>0.125</td></tr>
 <tr><td>Leacock-Chodorow Similarity</td><td>1.440361582</td><td>1.072636802</td><td>2.538973871</td><td>1.335001067</td><td>1.558144618</td></tr>
 <tr><td>Wu-Palmer Similarity</td><td>0.428571428</td><td>0.1428571429</td><td>0.8</td><td>0.3076923077</td><td>0.2222222222</td></tr>
 <tr><td>Resnik Similarity</td><td>2.619643841</td><td>0</td><td>1.531833743</td><td>0.5962292079</td><td>0</td></tr>
 <tr><td>Jiang-Conrath Similarity</td><td>0.071250949</td><td>0.1192745355</td><td>1.444425592</td><td>0.05928363701</td><td>0.07444639262</td></tr>
 <tr><td>Lin Similarity</td><td>0.271829224</td><td>0</td><td>0.8156763478</td><td>0.06602569917</td><td>0</td></tr>
</tbody></table>


Как можно заметить, по всем индексам наиболее похожими получились значения организма (a living thing that has (or can develop) the ability to act or function independently) и целого (an assemblage of parts that is regarded as a single entity). 

Как мне кажется, они и правда являются довольно близкими, даже ближе, чем синонимы спектакль и комедия так как в этом случае, комедия всё же имеет более узкое значение. 

Смущает предпоследний индекс Jiang-Conrath Similarity, потому что в нем у пары спектакля и комедии значение близкое к нулю, что противоречит интуиции.
